### PR TITLE
properly lock the target file on dav upload when not using part files

### PR DESCRIPTION
**Pull request by icewind1991**
While non-partfile storages don't have the file corruption problem on current read/writes (which is why I left this until now) not locking the target file during write can still lead to an inconsistent filecache state because the scanner is not being blocked.

### CODE REVIEW

File: apps/dav/lib/Connector/Sabre/File.php [hunk 1]
```diff
@@ -191,26 +200,26 @@ public function put($data) {
...
+				try {
+					$this->changeLock(ILockingProvider::LOCK_EXCLUSIVE);
+				} catch (LockedException $e) {
+					if ($needsPartFile) {
+						$partStorage->unlink($internalPartPath);
+					}
+					throw new FileLocked($e->getMessage(), $e->getCode(), $e);
 				}
-				throw new FileLocked($e->getMessage(), $e->getCode(), $e);
 			}
```
Comment (rullzer) [inline apps/dav/lib/Connector/Sabre/File.php]: No need for two ifs ;)

Comment (MorrisJobke) [reply apps/dav/lib/Connector/Sabre/File.php]: Why? IMO it makes sense to split it. 🤔

Comment (icewind1991) [reply apps/dav/lib/Connector/Sabre/File.php]: fixed

**Review by rullzer — APPROVED** [review]
Makes sense!

**Review by MorrisJobke — APPROVED** [review]

**Comment by MorrisJobke** [comment]
Unit tests fail. :/

File: apps/dav/lib/Connector/Sabre/File.php [hunk 2]
```diff
@@ -191,26 +200,24 @@ public function put($data) {
...
-			$run = ($view && $needsPartFile) ? $this->emitPreHooks($exists) : true;
-
-			try {
-				$this->changeLock(ILockingProvider::LOCK_EXCLUSIVE);
-			} catch (LockedException $e) {
-				if ($needsPartFile) {
-					$partStorage->unlink($internalPartPath);
+			if ($needsPartFile) {
+				if ($view) {
+					$this->emitPreHooks($exists);
```
Comment (rullzer) [inline apps/dav/lib/Connector/Sabre/File.php]: You ignore the result here now. So if the prehook returns false you should probably also fail right?

Comment (MorrisJobke) [reply apps/dav/lib/Connector/Sabre/File.php]: See the unit tests, which fail now due to this.

**Review by MorrisJobke — CHANGES_REQUESTED** [review]
See unit tests and @rullzer's comment.

**Comment by icewind1991** [comment]
Fixed the unit test

**Review by MorrisJobke — APPROVED** [review]

**Comment by MorrisJobke** [comment]
@icewind1991 Does a backport to 13 makes sense or should we skip it?

**Comment by icewind1991** [comment]
backport makes sense imo

**Comment by MorrisJobke** [comment]
@icewind1991 Could you prepare the PR?

**Pull request merged**